KAFKA-7890: Invalidate ClusterConnectionState cache for a broker if the hostname of the broker changes.

markcho opened the pull request:
See https://issues.apache.org/jira/browse/KAFKA-7890 for additional details.

- Added a new unit test where the hostname changes.
- Also tested in production

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

markcho committed:
KAFKA-7890: Invalidate ClusterConnectionState cache for a broker if the hostname of the broker changes.

cmccabe commented:
This looks reasonable.  Can you log a message when the hostname changes?  It should be very infrequent.

markcho committed:
Add info log for when a broker hostname changes.

markcho commented:
Thanks for taking a look @cmccabe.

Added log.info for broker hostname changing.

markcho committed:
Remove outdated test for hostname changing in ClusterConnectionStatesTest.

Previous test was using reflection to change the hostname field in the connectionStates map. This test is replaced with `testNodeWithNewHostname()`.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/ClusterConnectionStates.java:
```diff
@@ -361,6 +370,10 @@ public NodeConnectionState(ConnectionState state, long lastConnectAttempt, long
             this.clientDnsLookup = clientDnsLookup;
         }
 
+        public String getHost() {
```
nit: We don't use the 'get` prefix for getters, just `host()` to be consistent.

markcho replied:
Done

rajinisivaram commented on clients/src/test/java/org/apache/kafka/clients/ClusterConnectionStatesTest.java:
```diff
@@ -279,22 +278,13 @@ public void testMultipleIPsWithUseAll() throws UnknownHostException {
-        Field hostField = nodeState.getClass().getDeclaredField("host");
-        hostField.setAccessible(true);
-        hostField.set(nodeState, "localhost");
-
```
I can see why this test was removed, but I think there is a missing test case now. We are testing host name change with the updated test `testNodeWithNewHostname` that verifies the change from this PR. But we don't have a test that verifies the additional lookup added under KAFKA-7755 which was being tested earlier with a hostname change? We need a test that fails if we removed the look up added under KAFKA-7755.

markcho replied:
I restored the existing test and just updated a single line to work with the new code.

rajinisivaram reviewed (commented):
@markcho Thanks for the PR, looks good. Left a couple of minor comments.

markcho committed:
Rename getter without get. Restore and update previous test case.

rajinisivaram commented:
retest this please

rajinisivaram reviewed (approved):
@markcho Thanks for the updates, LGTM.

The test failures don't look related, but retrying the build once to see if we can get a clean build before merging. Will merge after builds complete.

rajinisivaram commented:
Test failure not related. Merging to trunk, 2.2 and 2.1.
